src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -897,25 +953,27 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
                     (remaining.Length == 0 ? FrameFlags.EndHeaders : FrameFlags.None) |
                     (request.Content == null ? FrameFlags.EndStream : FrameFlags.None);
 
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + current.Length);
+                await StartWriteAsync(FrameHeader.Size + current.Length, default, false).ConfigureAwait(false);
+
                 WriteFrameHeader(new FrameHeader(current.Length, FrameType.Headers, flags, streamId));
                 current.CopyTo(_outgoingBuffer.AvailableMemory);
                 _outgoingBuffer.Commit(current.Length);
 
-                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                await FinishWriteAsync(remaining.Length == 0, cancellationToken, false).ConfigureAwait(false);
 
                 while (remaining.Length > 0)
                 {
                     (current, remaining) = SplitBuffer(remaining, FrameHeader.MaxLength);
 
                     flags = (remaining.Length == 0 ? FrameFlags.EndHeaders : FrameFlags.None);
 
-                    _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + current.Length);
+                    await StartWriteAsync(FrameHeader.Size + current.Length, default, false).ConfigureAwait(false);
```

Comment:
I think we are still not handling write atomicity properly here. We could send a HEADERS block, then cancel before sending the CONTINUATION.

My suggestion here is to just write this all out in a single StartWriteAsync pass. I think that will also remove the need for the acquireWriteLock flag on StartWriteAsync, which will simplify things a bit.

This means we need a separate semaphore to serialize stream creation, but that seems fine. There's already a comment above saying we may want to do this, and it will probably reduce write lock contention in some scenarios.

Reply:
> My suggestion here is to just write this all out in a single StartWriteAsync pass. I think that will also remove the need for the acquireWriteLock flag on StartWriteAsync, which will simplify things a bit.

The code change here is simple enough, but it does mean that the write buffer can grow larger than the maximum header frame size in the case of super large headers. That's a very unusual case though, so I'm not sure we need to care about it.

Reply:
Yeah, I don't think we need to worry about that.